{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0a937f58",
   "metadata": {},
   "source": [
    "# Relative expression across cell types\n",
    "\n",
    "Let the matrix of mean expression values for each gene g in each cell type $ct$ for a given modality $m$ defined as:\n",
    "$$<X>_{ct,g}^m = \\frac{1}{\\lvert C_{ct} \\rvert} \\sum\\limits_{g\\in C_{ct}} X_{g,ct}^{m}$$ \n",
    "\n",
    "where $C_{ct}$ is the set of all genes in given cell type clusters $ct$ and $X_{c,g}^{m}$ is the matrix of normalized expression values for each gene $g$ in each cell $c$. \n",
    "\n",
    "Define the difference in mean expression between two cell types $(c_1,c_2)$ in the given gene $g$ and the given modality\n",
    "$$\\delta_{c_1,c_2}^m= <X>_{g,c_1}^m - <X>_{g,c_2}^m $$\n",
    "\n",
    "By deriving the mean expressions to the sum of the absolute values of all possible pairwise differences between cell types, we normalize sothat they are comparable across modalities. To be sure that the values are comparable across datasets with different numbers of cell-types, we scale the result by a factor of $N_{c}^2$ representing the total number of shared cell types shared between the two modalities. We define the normalized pairwise difference in mean expression between two cell types $c_1,c_2$ in a given gene $g$ and a given modality $m$ as\n",
    "$$\\delta_{c_1,c_2}^{'m,g}= \\frac{N_{c}^2 \\delta_{c_1,c_2}^{m,g}}{\\sum\\limits_{c_1}\\sum\\limits_{c_2}|\\delta_{c_1,c_2}^{m,g}|} $$\n",
    "\n",
    "To compare the normalized pairwise differences in mean cell-type expression between modalities, we define $\\triangle$ as\n",
    "$$\\triangle = \\sum\\limits_{ct}\\sum\\limits_{c_1}\\sum\\limits_{c_2} |\\delta_{c_1,c_2}^{'sp,g}- \\delta_{c_1,c_2}^{'sc,g}|$$\n",
    "\n",
    "Define the final metric $M$ bounded at a maximum of 1, representing the perfect similarity of relative gene expression between modalities, and minimum of 0, representing the perfect dissimilarity of relative cell type expression between two modalities such that each cell types expression value in each gene pair is swapped.\n",
    "\n",
    "$$M=1-\\frac{\\triangle}{2*|\\sum\\limits_{g,c_1,c_1}\\delta_{c_1,c_2}^{'sc,g}|}$$\n",
    "\n",
    "Further compute the metric on per-gene and per-celltype basis:\n",
    "\n",
    "1. per cell-type:\n",
    "\n",
    "$$\\triangle_{c1}= \\sum\\limits_{g}\\sum\\limits_{c_2}|\\delta_{c_1,c_2}^{'sp,g}- \\delta_{c_1,c_2}^{'sc,g}|$$\n",
    "$$M_{c_1}=1-\\frac{\\triangle}{2*|\\sum\\limits_{g,c_2}\\delta_{c_1,c_2}^{'sc,g}|}$$\n",
    "\n",
    "2. per gene:\n",
    "$$\\triangle_{g}= \\sum\\limits_{c_1}\\sum\\limits_{c_2}|\\delta_{c_1,c_2}^{'sp,g}- \\delta_{c_1,c_2}^{'sc,g}|$$\n",
    "$$M_{g}=1-\\frac{\\triangle}{2*|\\sum\\limits_{c_1,c_2}\\delta_{c_1,c_2}^{'sc,g}|}$$\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aae4bffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from anndata import AnnData\n",
    "from scipy.sparse import issparse\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "c222c0c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/miniconda3/envs/project-4-env/lib/python3.10/site-packages/scanpy/preprocessing/_normalization.py:197: UserWarning: Some cells have zero counts\n",
      "  warn(UserWarning('Some cells have zero counts'))\n"
     ]
    }
   ],
   "source": [
    "adata_sc = sc.read_h5ad(\"/mnt/storage/adata_sc.h5ad\")\n",
    "adata_sc.layers[\"raw\"] = adata_sc.X.copy()\n",
    "sc.pp.normalize_total(adata_sc)\n",
    "adata_sc.layers[\"norm\"] = adata_sc.X.copy()\n",
    "sc.pp.log1p(adata_sc)\n",
    "adata_sc.layers[\"lognorm\"] = adata_sc.X.copy()\n",
    "adata_sp = sc.read_h5ad(\"/mnt/storage/adata_sp.h5ad\")\n",
    "adata_sp.layers[\"raw\"] = adata_sp.X.copy()\n",
    "sc.pp.normalize_total(adata_sp)\n",
    "adata_sp.layers[\"norm\"] = adata_sp.X.copy()\n",
    "sc.pp.log1p(adata_sp)\n",
    "adata_sp.layers[\"lognorm\"] = adata_sp.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "49de7632",
   "metadata": {},
   "outputs": [],
   "source": [
    "  def relative_celltype_expression(adata_sp: AnnData, adata_sc: AnnData, key:str='celltype', layer:str='lognorm'):\n",
    "    \"\"\"Calculate the efficiency deviation present between the genes in the panel. \n",
    "    ----------\n",
    "    adata_sp : AnnData\n",
    "        annotated ``AnnData`` object with counts from spatial data\n",
    "    adata_sc : AnnData\n",
    "        annotated ``AnnData`` object with counts from scRNAseq data\n",
    "    key: str (default: 'celltype')\n",
    "        .obs column of ``AnnData`` that contains celltype information\n",
    "    layer: str (default: 'lognorm')\n",
    "        layer of ```AnnData`` to use to compute the metric\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    overall_metric: float\n",
    "        similarity of relative gene expression across all genes and celltypes, b/t the scRNAseq and spatial data\n",
    "    per_gene_metric: float\n",
    "        similarity of relative gene expression per gene across all celltypes, b/t the scRNAseq and spatial data\n",
    "    per_celltype_metric: float\n",
    "        similarity of relative gene expression per celltype across all genes, b/t the scRNAseq and spatial data\n",
    "  \n",
    "    \"\"\"   \n",
    "    ### SET UP\n",
    "    # set the .X layer of each of the adatas to be log-normalized counts\n",
    "    adata_sp.X = adata_sp.layers[layer]\n",
    "    adata_sc.X = adata_sc.layers[layer]\n",
    "    \n",
    "    # take the intersection of genes in adata_sp and adata_sc, as a list\n",
    "    intersect = list(set(adata_sp.var_names).intersection(set(adata_sc.var_names)))\n",
    "    \n",
    "    # subset adata_sc and adata_sp to only include genes in the intersection of adata_sp and adata_sc \n",
    "    adata_sc=adata_sc[:,intersect]\n",
    "    adata_sp=adata_sp[:,intersect]\n",
    "    \n",
    "    # sparse matrix support\n",
    "    for a in [adata_sc, adata_sp]:\n",
    "        if issparse(a.X):\n",
    "            a.X = a.X.toarray()\n",
    "            \n",
    "    # find the unique celltypes in adata_sc that are also in adata_sp\n",
    "    unique_celltypes=adata_sc.obs.loc[adata_sc.obs[key].isin(adata_sp.obs[key]),key].unique()\n",
    "    \n",
    "    \n",
    "    \n",
    "    #### FIND MEAN GENE EXPRESSION PER CELL TYPE FOR EACH MODALITY\n",
    "    # get the adata_sc cell x gene matrix as a pandas dataframe (w gene names as column names)\n",
    "    exp_sc=pd.DataFrame(adata_sc.X,columns=adata_sc.var.index)\n",
    "    \n",
    "    # get the adata_sp cell x gene matrix as a pandas dataframe (w gene names as column names)\n",
    "    exp_sp=pd.DataFrame(adata_sp.X,columns=adata_sp.var.index)\n",
    "    \n",
    "    # add \"celltype\" label column to exp_sc & exp_sp cell x gene matrices \n",
    "    exp_sc[key]=list(adata_sc.obs[key])\n",
    "    exp_sp[key]=list(adata_sp.obs[key])\n",
    "    \n",
    "    # delete all cells from the exp matrices if they aren't in the set of intersecting celltypes b/t sc & sp data\n",
    "    exp_sc=exp_sc.loc[exp_sc[key].isin(unique_celltypes),:]\n",
    "    exp_sp=exp_sp.loc[exp_sp[key].isin(unique_celltypes),:]\n",
    "    \n",
    "    # find the mean expression for each gene for each celltype in sc and sp data\n",
    "    mean_celltype_sp=exp_sp.groupby(key).mean()\n",
    "    mean_celltype_sc=exp_sc.groupby(key).mean()\n",
    "    \n",
    "    # sort genes in alphabetical order \n",
    "    mean_celltype_sc=mean_celltype_sc.loc[:,mean_celltype_sc.columns.sort_values()]\n",
    "    mean_celltype_sp=mean_celltype_sp.loc[:,mean_celltype_sp.columns.sort_values()]\n",
    "    \n",
    "    \n",
    "    #### CALCULATE PAIRWISE RELATIVE DISTANCES BETWEEN CELL TYPES\n",
    "    mean_celltype_sc_np = mean_celltype_sc.T.to_numpy()\n",
    "    pairwise_distances_sc = mean_celltype_sc_np[:,:,np.newaxis] - mean_celltype_sc_np[:,np.newaxis,:]\n",
    "    pairwise_distances_sc = pairwise_distances_sc.transpose((1,2,0)) #results in np.array of dimensions (num_genes, num_genes, num_celltypes) \n",
    "       \n",
    "    mean_celltype_sp_np = mean_celltype_sp.T.to_numpy()\n",
    "    pairwise_distances_sp = mean_celltype_sp_np[:,:,np.newaxis] - mean_celltype_sp_np[:,np.newaxis,:]\n",
    "    pairwise_distances_sp = pairwise_distances_sp.transpose((1,2,0)) #results in np.array of dimensions (num_genes, num_genes, num_celltypes) \n",
    "    \n",
    "    #### NORMALIZE THESE PAIRWISE DISTANCES BETWEEN CELL TYPES\n",
    "    #calculate sum of absolute distances\n",
    "    abs_diff_sc = np.absolute(pairwise_distances_sc.T)\n",
    "    abs_diff_sum_sc = np.sum(abs_diff_sc, axis=(0,1))\n",
    "    \n",
    "    abs_diff_sp = np.absolute(pairwise_distances_sp.T)\n",
    "    abs_diff_sum_sp = np.sum(abs_diff_sp, axis=(0,1))\n",
    "    \n",
    "    norm_factor_sc = mean_celltype_sc.T.shape[1]**2 * abs_diff_sum_sc\n",
    "    norm_factor_sp = mean_celltype_sp.T.shape[1]**2 * abs_diff_sum_sp\n",
    "    \n",
    "    #perform normalization\n",
    "    norm_pairwise_distances_sc = np.divide(pairwise_distances_sc.T, norm_factor_sc)\n",
    "    norm_pairwise_distances_sp = np.divide(pairwise_distances_sp.T, norm_factor_sp)\n",
    "    \n",
    "    ##### CALCULATE OVERALL SCORE,PER-GENE SCORES, PER-CELLTYPE SCORES\n",
    "    overall_score = np.sum(np.absolute(norm_pairwise_distances_sp - norm_pairwise_distances_sc), axis=None)\n",
    "    overall_metric = 1 - (overall_score/(2 * np.sum(np.absolute(norm_pairwise_distances_sc), axis=None)))\n",
    "    \n",
    "    per_gene_score = np.sum(np.absolute(norm_pairwise_distances_sp - norm_pairwise_distances_sc), axis=(1,2))\n",
    "    per_gene_metric = 1 - (per_gene_score/(2 * np.sum(np.absolute(norm_pairwise_distances_sc), axis=(1,2))))\n",
    "    per_gene_metric = pd.DataFrame(per_gene_metric, index=mean_celltype_sc.columns, columns=['score']) #add back the gene labels \n",
    "\n",
    "\n",
    "    #per_gene_metric = pd.DataFrame(per_gene_metric, index=mean_celltype_sc.T.columns, columns=['score']) #add back the gene labels \n",
    "    \n",
    "    per_celltype_score = np.sum(np.absolute(norm_pairwise_distances_sp - norm_pairwise_distances_sc), axis=(0,1))\n",
    "    per_celltype_metric = 1 - (per_celltype_score/(2 * np.sum(np.absolute(norm_pairwise_distances_sc), axis=(0,1))))\n",
    "    per_celltype_metric = pd.DataFrame(per_celltype_metric, index=mean_celltype_sc.index, columns=['score']) #add back the celltype labels \n",
    "    \n",
    "    return overall_metric, per_gene_metric, per_celltype_metric\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "899c093a",
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_metric, per_gene_metric, per_celltype_metric = relative_celltype_expression(adata_sp, adata_sc,'celltype', 'lognorm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "dd21c5b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.58707479695175"
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overall_metric\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "457cec09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALDH1A1</th>\n",
       "      <td>-2.452862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC102B</th>\n",
       "      <td>0.759614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDK1</th>\n",
       "      <td>0.538605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLDN5</th>\n",
       "      <td>0.553949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLU</th>\n",
       "      <td>0.505288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNNI1</th>\n",
       "      <td>0.636490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNNT1</th>\n",
       "      <td>0.530408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOP2A</th>\n",
       "      <td>0.182024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPM1</th>\n",
       "      <td>0.574190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIL</th>\n",
       "      <td>0.636117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             score\n",
       "ALDH1A1  -2.452862\n",
       "CCDC102B  0.759614\n",
       "CDK1      0.538605\n",
       "CLDN5     0.553949\n",
       "CLU       0.505288\n",
       "...            ...\n",
       "TNNI1     0.636490\n",
       "TNNT1     0.530408\n",
       "TOP2A     0.182024\n",
       "TPM1      0.574190\n",
       "TRIL      0.636117\n",
       "\n",
       "[66 rows x 1 columns]"
      ]
     },
     "execution_count": 311,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The history saving thread hit an unexpected error (OperationalError('database or disk is full')).History will not be written to the database.\n"
     ]
    }
   ],
   "source": [
    "per_gene_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "e2f0d387",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>STMN2</th>\n",
       "      <td>-23.574125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCSK1N</th>\n",
       "      <td>-18.955132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISL1</th>\n",
       "      <td>-5.071395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDH1A1</th>\n",
       "      <td>-2.452862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYRF</th>\n",
       "      <td>-2.364538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL3</th>\n",
       "      <td>0.798107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A1</th>\n",
       "      <td>0.808010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAV1</th>\n",
       "      <td>0.832547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LDB3</th>\n",
       "      <td>0.863535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM100</th>\n",
       "      <td>0.905663</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             score\n",
       "STMN2   -23.574125\n",
       "PCSK1N  -18.955132\n",
       "ISL1     -5.071395\n",
       "ALDH1A1  -2.452862\n",
       "MYRF     -2.364538\n",
       "...            ...\n",
       "MYL3      0.798107\n",
       "COL1A1    0.808010\n",
       "NAV1      0.832547\n",
       "LDB3      0.863535\n",
       "TMEM100   0.905663\n",
       "\n",
       "[66 rows x 1 columns]"
      ]
     },
     "execution_count": 312,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_gene_metric.sort_values(by = 'score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "d75427ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Atrial cardiomyocytes</th>\n",
       "      <td>0.627987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Capillary endothelium</th>\n",
       "      <td>0.576846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cardiac neural crest cells</th>\n",
       "      <td>0.451244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endothelium / pericytes</th>\n",
       "      <td>0.487327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epicardial cells</th>\n",
       "      <td>0.502080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroblast-like</th>\n",
       "      <td>0.641672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Myoz2-enriched cardiomyocytes</th>\n",
       "      <td>0.641890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smooth muscle cells</th>\n",
       "      <td>0.683424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ventricular cardiomyocytes</th>\n",
       "      <td>0.671203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  score\n",
       "celltype                               \n",
       "Atrial cardiomyocytes          0.627987\n",
       "Capillary endothelium          0.576846\n",
       "Cardiac neural crest cells     0.451244\n",
       "Endothelium / pericytes        0.487327\n",
       "Epicardial cells               0.502080\n",
       "Fibroblast-like                0.641672\n",
       "Myoz2-enriched cardiomyocytes  0.641890\n",
       "Smooth muscle cells            0.683424\n",
       "Ventricular cardiomyocytes     0.671203"
      ]
     },
     "execution_count": 313,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_celltype_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "e9af74a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "    ### SET UP\n",
    "    # set the .X layer of each of the adatas to be log-normalized counts\n",
    "    adata_sp.X = adata_sp.layers['lognorm']\n",
    "    adata_sc.X = adata_sc.layers['lognorm']\n",
    "    \n",
    "    # take the intersection of genes in adata_sp and adata_sc, as a list\n",
    "    intersect = list(set(adata_sp.var_names).intersection(set(adata_sc.var_names)))\n",
    "    \n",
    "    # subset adata_sc and adata_sp to only include genes in the intersection of adata_sp and adata_sc \n",
    "    adata_sc=adata_sc[:,intersect]\n",
    "    adata_sp=adata_sp[:,intersect]\n",
    "    \n",
    "    # sparse matrix support\n",
    "    for a in [adata_sc, adata_sp]:\n",
    "        if issparse(a.X):\n",
    "            a.X = a.X.toarray()\n",
    "            \n",
    "    # find the unique celltypes in adata_sc that are also in adata_sp\n",
    "    unique_celltypes=adata_sc.obs.loc[adata_sc.obs[key].isin(adata_sp.obs[key]),key].unique()\n",
    "    \n",
    "    \n",
    "    \n",
    "    #### FIND MEAN GENE EXPRESSION PER CELL TYPE FOR EACH MODALITY\n",
    "    # get the adata_sc cell x gene matrix as a pandas dataframe (w gene names as column names)\n",
    "    exp_sc=pd.DataFrame(adata_sc.X,columns=adata_sc.var.index)\n",
    "    \n",
    "    # get the adata_sp cell x gene matrix as a pandas dataframe (w gene names as column names)\n",
    "    exp_sp=pd.DataFrame(adata_sp.X,columns=adata_sp.var.index)\n",
    "    \n",
    "    # add \"celltype\" label column to exp_sc & exp_sp cell x gene matrices \n",
    "    exp_sc[key]=list(adata_sc.obs[key])\n",
    "    exp_sp[key]=list(adata_sp.obs[key])\n",
    "    \n",
    "    # delete all cells from the exp matrices if they aren't in the set of intersecting celltypes b/t sc & sp data\n",
    "    exp_sc=exp_sc.loc[exp_sc[key].isin(unique_celltypes),:]\n",
    "    exp_sp=exp_sp.loc[exp_sp[key].isin(unique_celltypes),:]\n",
    "    \n",
    "    # find the mean expression for each gene for each celltype in sc and sp data\n",
    "    mean_celltype_sp=exp_sp.groupby(key).mean()\n",
    "    mean_celltype_sc=exp_sc.groupby(key).mean()\n",
    "    \n",
    "    # sort genes in alphabetical order \n",
    "    mean_celltype_sc=mean_celltype_sc.loc[:,mean_celltype_sc.columns.sort_values()]\n",
    "    mean_celltype_sp=mean_celltype_sp.loc[:,mean_celltype_sp.columns.sort_values()]\n",
    "    \n",
    "    \n",
    "    #### CALCULATE PAIRWISE RELATIVE DISTANCES BETWEEN CELL TYPES\n",
    "    mean_celltype_sc_np = mean_celltype_sc.T.to_numpy()\n",
    "    pairwise_distances_sc = mean_celltype_sc_np[:,:,np.newaxis] - mean_celltype_sc_np[:,np.newaxis,:]\n",
    "    pairwise_distances_sc = pairwise_distances_sc.transpose((1,2,0)) #results in np.array of dimensions (num_celltypes, num_celltypes, num_genes) \n",
    "       \n",
    "    mean_celltype_sp_np = mean_celltype_sp.T.to_numpy()\n",
    "    pairwise_distances_sp = mean_celltype_sp_np[:,:,np.newaxis] - mean_celltype_sp_np[:,np.newaxis,:]\n",
    "    pairwise_distances_sp = pairwise_distances_sp.transpose((1,2,0)) #results in np.array of dimensions (num_celltypes, num_celltypes, num_genes) \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "95ebda6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "    #### normalize these pairwise distances between cell types\n",
    "    #calculate sum of absolute distances\n",
    "    abs_diff_sc = np.absolute(pairwise_distances_sc.T)\n",
    "    abs_diff_sum_sc = np.sum(abs_diff_sc, axis=(0,1))\n",
    "    \n",
    "    abs_diff_sp = np.absolute(pairwise_distances_sp.T)\n",
    "    abs_diff_sum_sp = np.sum(abs_diff_sp, axis=(0,1))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "72ab8661",
   "metadata": {},
   "outputs": [],
   "source": [
    "    # calculate normalization factor\n",
    "    norm_factor_sc = mean_celltype_sc.T.shape[1]**2 * abs_diff_sum_sc\n",
    "    norm_factor_sp = mean_celltype_sp.T.shape[1]**2 * abs_diff_sum_sp\n",
    "    \n",
    "    #perform normalization\n",
    "    norm_pairwise_distances_sc = np.divide(pairwise_distances_sc.T, norm_factor_sc)\n",
    "    norm_pairwise_distances_sp = np.divide(pairwise_distances_sp.T, norm_factor_sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "655b511d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.58707479695175"
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "    ##### CALCULATE OVERALL SCORE,PER-GENE SCORES, PER-CELLTYPE SCORES\n",
    "    overall_score = np.sum(np.absolute(norm_pairwise_distances_sp - norm_pairwise_distances_sc), axis=None)\n",
    "    overall_metric = 1 - (overall_score/(2 * np.sum(np.absolute(norm_pairwise_distances_sc), axis=None)))\n",
    "    \n",
    "    per_gene_score = np.sum(np.absolute(norm_pairwise_distances_sp - norm_pairwise_distances_sc), axis=(1,2))\n",
    "    per_gene_metric = 1 - (per_gene_score/(2 * np.sum(np.absolute(norm_pairwise_distances_sc), axis=(1,2))))\n",
    "    per_gene_metric = pd.DataFrame(per_gene_metric, index=mean_celltype_sc.columns, columns=['score']) #add back the gene labels \n",
    "    \n",
    "    overall_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "cd5bd653",
   "metadata": {},
   "outputs": [],
   "source": [
    "    per_celltype_score = np.sum(np.absolute(norm_pairwise_distances_sp - norm_pairwise_distances_sc), axis=(0,1))\n",
    "    per_celltype_metric = 1 - (per_celltype_score/(2 * np.sum(np.absolute(norm_pairwise_distances_sc), axis=(0,1))))\n",
    "    per_celltype_metric = pd.DataFrame(per_celltype_metric, index=mean_celltype_sc.index, columns=['score']) #add back the celltype labels \n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "9b30328c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALDH1A1</th>\n",
       "      <td>-2.452862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC102B</th>\n",
       "      <td>0.759614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDK1</th>\n",
       "      <td>0.538605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLDN5</th>\n",
       "      <td>0.553949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLU</th>\n",
       "      <td>0.505288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNNI1</th>\n",
       "      <td>0.636490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNNT1</th>\n",
       "      <td>0.530408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOP2A</th>\n",
       "      <td>0.182024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPM1</th>\n",
       "      <td>0.574190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIL</th>\n",
       "      <td>0.636117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             score\n",
       "ALDH1A1  -2.452862\n",
       "CCDC102B  0.759614\n",
       "CDK1      0.538605\n",
       "CLDN5     0.553949\n",
       "CLU       0.505288\n",
       "...            ...\n",
       "TNNI1     0.636490\n",
       "TNNT1     0.530408\n",
       "TOP2A     0.182024\n",
       "TPM1      0.574190\n",
       "TRIL      0.636117\n",
       "\n",
       "[66 rows x 1 columns]"
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_gene_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "93c69b17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>STMN2</th>\n",
       "      <td>-23.574125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCSK1N</th>\n",
       "      <td>-18.955132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISL1</th>\n",
       "      <td>-5.071395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDH1A1</th>\n",
       "      <td>-2.452862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYRF</th>\n",
       "      <td>-2.364538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL3</th>\n",
       "      <td>0.798107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A1</th>\n",
       "      <td>0.808010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAV1</th>\n",
       "      <td>0.832547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LDB3</th>\n",
       "      <td>0.863535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM100</th>\n",
       "      <td>0.905663</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             score\n",
       "STMN2   -23.574125\n",
       "PCSK1N  -18.955132\n",
       "ISL1     -5.071395\n",
       "ALDH1A1  -2.452862\n",
       "MYRF     -2.364538\n",
       "...            ...\n",
       "MYL3      0.798107\n",
       "COL1A1    0.808010\n",
       "NAV1      0.832547\n",
       "LDB3      0.863535\n",
       "TMEM100   0.905663\n",
       "\n",
       "[66 rows x 1 columns]"
      ]
     },
     "execution_count": 302,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_gene_metric.sort_values(by = 'score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "ddc2eea6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Atrial cardiomyocytes</th>\n",
       "      <td>0.627987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Capillary endothelium</th>\n",
       "      <td>0.576846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cardiac neural crest cells</th>\n",
       "      <td>0.451244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endothelium / pericytes</th>\n",
       "      <td>0.487327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epicardial cells</th>\n",
       "      <td>0.502080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroblast-like</th>\n",
       "      <td>0.641672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Myoz2-enriched cardiomyocytes</th>\n",
       "      <td>0.641890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smooth muscle cells</th>\n",
       "      <td>0.683424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ventricular cardiomyocytes</th>\n",
       "      <td>0.671203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  score\n",
       "celltype                               \n",
       "Atrial cardiomyocytes          0.627987\n",
       "Capillary endothelium          0.576846\n",
       "Cardiac neural crest cells     0.451244\n",
       "Endothelium / pericytes        0.487327\n",
       "Epicardial cells               0.502080\n",
       "Fibroblast-like                0.641672\n",
       "Myoz2-enriched cardiomyocytes  0.641890\n",
       "Smooth muscle cells            0.683424\n",
       "Ventricular cardiomyocytes     0.671203"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_celltype_metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d7df250",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
